File: c/easy/count_binary_substrings.c
#include <stdio.h>
#include <string.h>

int countBinarySubstrings(const char *s) {
    int total_substrings = 0;
    int previous_run_length = 0;
    int current_run_length = 1;
    int len = strlen(s);

    for (int i = 1; i < len; i++) {
        if (s[i] == s[i - 1]) {
            current_run_length++;
        } else {
            previous_run_length = current_run_length;
            current_run_length = 1;
        }

        if (current_run_length <= previous_run_length) {
            total_substrings++;
        }
    }

    return total_substrings;
}

File: rust/easy/count_binary_substrings.rs
pub fn count_binary_substrings(s: String) -> i32 {
    let bytes = s.as_bytes();
    let mut total_substrings = 0;
    let mut previous_run_length = 0;
    let mut current_run_length = 1;

    for i in 1..bytes.len() {
        if bytes[i] == bytes[i - 1] {
            current_run_length += 1;
        } else {
            previous_run_length = current_run_length;
            current_run_length = 1;
        }

        if current_run_length <= previous_run_length {
            total_substrings += 1;
        }
    }

    total_substrings
}
